spec/index.bs

```diff
@@ -8459,58 +8468,67 @@ clip space positions for [[#primitive-clipping]], as well as other data for the
-        1. Set the shader builtins:
-            - Set the `VertexIndex` builtin, if any, to |vertexIndex|.
-            - Set the `InstanceIndex` builtin, if any, to |instanceIndex|.
-        1. Invoke vertex shader entry point described by |desc|.
-
-            Note: The target platform caches the results of vertex shader invocations.
-            There is no guarantee that any |vertexIndex| that repeats more than once will
-            result in multiple invocations. Similarly, there is no guarantee that a single |vertexIndex|
-            will only be processed once.
+    **Arguments:**
+    - |vertexIndexList|: List of vertex indices to process (mutable, passed by reference).
+    - |drawCall|: The draw call parameters.
+    - |desc|: The descriptor of type {{GPUVertexState}}.
+    - |state|: The active [=RenderState=].
+
+    Each vertex |vertexIndex| in the |vertexIndexList|,
+    in each instance of index |rawInstanceIndex|, is processed independently.
+    The |rawInstanceIndex| is in range from 0 to |drawCall|.instanceCount - 1, inclusive.
+    This processing happens in parallel, and any side effects, such as
+    writes into {{GPUBufferBindingType/"storage"|GPUBufferBindingType."storage"}} bindings,
+    may happen in any order.
+    1. Let |instanceIndex| be |rawInstanceIndex| + |drawCall|.baseInstance.
+    1. For each non-`null` |vertexBufferLayout| in the list of |desc|.{{GPUVertexState/buffers}}:
+        1. Let |i| be the index of the buffer layout in this list.
+        1. Let |vertexBuffer| and |vertexBufferOffset| be the buffer and offset in
+            |state|.[=RenderState/vertexBuffers=] bindings at slot |i|
+        1. Let |vertexElementIndex| be dependent on |vertexBufferLayout|.{{GPUVertexBufferLayout/stepMode}}:
+            <dl class="switch">
+                : {{GPUInputStepMode/"vertex"}}
+                :: |vertexIndex|
+                : {{GPUInputStepMode/"instance"}}
+                :: |instanceIndex|
+            </dl>
+        1. For each |attributeDesc| in |vertexBufferLayout|.{{GPUVertexBufferLayout/attributes}}:
+            1. Let |attributeOffset| be |vertexBufferOffset| +
+                |vertexElementIndex| * |vertexBufferLayout|.{{GPUVertexBufferLayout/arrayStride}} +
+                |attributeDesc|.{{GPUVertexAttribute/offset}}.
+            1. Load the attribute |data| of format |attributeDesc|.{{GPUVertexAttribute/format}}
+                from |vertexBuffer| starting at offset |attributeOffset|.
+                If this results in an out-of-bounds access, the resulting value is determined
+                according to WGSL's [=invalid memory reference=] behavior.
+            1. **Optionally (implementation-defined):**
+                If |attributeOffset| + sizeof(|attributeDesc|.{{GPUVertexAttribute/format}} &gt;
+                |vertexBuffer|.{{GPUBuffer/[[size]]}},
```

Comment:
we should be comparing against the active range specified to `setVertexBuffer`, not to the whole size